internal/errorlint/errorlint.go

```diff
@@ -0,0 +1,191 @@
+
+	errors.New("Org not found in LAUNCHDARKLY rule")
+An invalid example is the following:
+
+	errors.New("Org not found in launchdarkly rule:")`
+
+// Analyzer exports the errorlint analyzer (linter).
+var Analyzer = analysis.Analyzer{
+	Name: name,
+	Doc:  doc,
+	Run:  errorlint,
+}
+
+// file represents a file being linted.
+type file struct {
+	f *ast.File
+}
+
+func (file *file) walk(fn func(ast.Node) bool) {
+	ast.Walk(walker(fn), file.f)
+}
+
+// walker adapts a function to satisfy the ast.Visitor interface.
+// The function returns whether the walk should proceed into the node's children.
+type walker func(ast.Node) bool
+
+func (w walker) Visit(node ast.Node) ast.Visitor {
+	if w(node) {
+		return w
+	}
+
+	return nil
+}
+
+// errorlint defines linter for error/trace/log messages
+func errorlint(_pass *analysis.Pass) (interface{}, error) {
+	// Ignore test packages.
+	if common.IsTestPackage(_pass) {
+		return nil, nil
+	}
+
+	// Wrap _pass with reporter.Pass to take nolint directives into account.
+	pass := reporter.NewPass(name, _pass, reporter.Warn())
+	for _, astFile := range pass.Files {
+		// Ignore generated files and test files.
+		if common.IsGenerated(astFile) || common.IsTestFile(pass.Pass, astFile) {
+			continue
+		}
+		newFile := &file{f: astFile}
+		lintMessageStrings(newFile, pass)
+	}
+
+	return nil, nil
+}
+
+// lintMessageStrings examines error/trace/log message strings for capitalization and valid ending
+func lintMessageStrings(file *file, pass *reporter.Pass) {
+	file.walk(func(node ast.Node) bool {
+		call, ok := node.(*ast.CallExpr)
+		if !ok {
+			return true
+		}
+
+		if isErrorPackage(call.Fun) || len(call.Args) < 1 {
+			return true
+		}
+
+		msgIndex := 1
+		if isDotInPkg(call.Fun, "errors", "New") || isDotInPkg(call.Fun, "fmt", "Errorf") ||
+			isDotInPkg(call.Fun, "errors", "Errorf") {
+			msgIndex = 0
+		}
+
+		msg, ok := call.Args[msgIndex].(*ast.BasicLit)
+		if !ok || msg.Kind != token.STRING {
+			return true
+		}
+
+		msgString, err := strconv.Unquote(msg.Value)
+		if err != nil {
+			return false
+		}
+
+		if msgString == "" {
+			return true
+		}
+		errormsg := getErrorMessage(msgString)
+		if errormsg != "" {
+			pkgName := getPkgName(call.Fun)
+			pass.Reportf(node.Pos(), "%s "+errormsg, pkgName)
+		}
+		return true
+	})
+}
+
+// isErrorPackage checks if the ast.Expr package matches the error/fmt/trace/log packages for linter
+func isErrorPackage(expr ast.Expr) bool {
+	return !isDotInPkg(expr, "errors", "New") && !isDotInPkg(expr, "errors", "Wrap") &&
+		!isDotInPkg(expr, "errors", "Wrapf") && !isDotInPkg(expr, "log", "Warn") &&
+		!isDotInPkg(expr, "log", "Info") && !isDotInPkg(expr, "log", "Error") &&
+		!isDotInPkg(expr, "trace", "StartSpan") && !isDotInPkg(expr, "trace", "StartCall") &&
+		!isDotInPkg(expr, "fmt", "Errorf") && !isDotInPkg(expr, "errors", "Errorf") &&
+		!isDotInPkg(expr, "errors", "WithMessage") && !isDotInPkg(expr, "errors", "WithMessagef")
+}
```

Comment:
This function doesn't take into account renamed packages. Example:

```go
package foo

import (
  _errors "errors"
)

func bar() error {
  return _errors.New("Invalid error string.")
}
```

Reply:
added this change, I wanted node.(*ast.CallExpr) and node.(*ast.ImportSpec) which are in different branches of ast GenDecl and FuncDecl, Type Assertions did not work here as there is no child to parent pointers here, so I created a file struct mapping each file to its import Declarations with file as key for lookup in map for renamed package info.